## Reading files

In [1]:
homer = open('homer_characters.txt','r').read()

In [2]:
homer[:100]

'Abarbarea         AB Abarbarea, Trojan fountain nymph\nAbas              19 Abas, brother of Polyeidu'

In [3]:
fh = open('homer_characters.txt','r')

In [4]:
fh

<open file 'homer_characters.txt', mode 'r' at 0x1023254b0>

## File stream iterator

In [5]:
next(fh)

'Abarbarea         AB Abarbarea, Trojan fountain nymph\n'

In [6]:
next(fh)

'Abas              19 Abas, brother of Polyeidus, killed by Diomedes\n'

In [5]:
fh = open('homer_characters.txt','r')
for line in fh:
    print(line.strip())

Abarbarea         AB Abarbarea, Trojan fountain nymph
Abas              19 Abas, brother of Polyeidus, killed by Diomedes
Ablerus           25 Ablerus, Trojan soldier killed by Antilochus
Acamas            02 Acamas, son of Antenor, lieutenant of Aeneas
Acamas_2          1U Acamas, Thracian commander, killed by Great
Achilles          AC Achilles, angry warrior, swift-footed chief of Myrmidons from Phthia
Actor             57 Actor, brother of Augeias
Adamas            3W Adamas, son of Asius, killed by Meriones
Adrestus          03 Adrestus, son of Merops, co-leader of Adresteians
Adrestus_2        A1 Adrestus, Trojan warrior killed by Patroclus
Aegaeon           AZ Aegaeon, hundred-armed giant, brother of Cronus
Aeneas            AE Aeneas, leader of Dardanians
Aenius            Q2 Aenius, charioteer from Paeonia killed by Achilles
Aeolus            WI Aeolus, lord of winds
Aesepus           1Z Aesepus, son of Bucolion and Abarbarea, killed by Euryalus
Aesymnus          71 Aesymnus, 

In [7]:
homer[:100]

'Abarbarea         AB Abarbarea, Trojan fountain nymph\nAbas              19 Abas, brother of Polyeidu'

In [2]:
import re
words = re.findall(r'\w+', open('homer_characters.txt').read().lower())

In [3]:
words[:10]

['abarbarea',
 'ab',
 'abarbarea',
 'trojan',
 'fountain',
 'nymph',
 'abas',
 '19',
 'abas',
 'brother']

In [4]:
words.sort()

In [5]:
words[:10]

['01', '02', '03', '04', '05', '06', '07', '08', '09', '0a']

In [6]:
words.sort(reverse=True)

In [13]:
words[:10]

['zf', 'zeus', 'zeus', 'zeus', 'zeus', 'zeus', 'zeus', 'zeus', 'zeus', 'zeus']

## NLTK (Natural Language Tool kit)

In [18]:
import nltk

In [ ]:
nltk.download()

## File streams

In [7]:
file_obj = open('pride_and_prejudice.txt','r')
file_obj

<open file 'pride_and_prejudice.txt', mode 'r' at 0x102e8f030>

In [9]:
ofh = open('foo.txt','w')
print('Great news!', file=ofh)
print("Monty Python lives!", file=ofh)
print('Hooray!', file=ofh)
ofh.close()

In [10]:
ifh = open('foo.txt','r')
line1 = ifh.readline()
print(line1)
line2 = ifh.readline()
print(line2)
line3 = ifh.readline()
print(line3)
line4 = ifh.readline()
line4

Great news!

Monty Python lives!

Hooray!



''

In [11]:
ifh = open('foo.txt','r')
line1 = next(ifh)
print(line1)
line2 = next(ifh)
print(line2)
line3 = next(ifh)
print(line3)

Great news!

Monty Python lives!

Hooray!



In [12]:
line4 = next(ifh)

StopIteration: 

## Reading *Pride and Prejudice*

In [13]:
from collections import Counter
ctr = Counter()
token_ctr = 0
with open('pride_and_prejudice.txt','r') as file_handle:
      for line in file_handle:
          line_words = line.strip().split()
          for word in line_words:
              token_ctr += 1
              ctr[word] += 1

We now discover

In [14]:
print len(ctr)
  print token_ctr

13638
124588


What happened?  The vocabulary size is 13,638; that's what we're counting:

In [15]:
ctr.most_common(10)

[('the', 4205),
 ('to', 4121),
 ('of', 3660),
 ('and', 3309),
 ('a', 1945),
 ('her', 1858),
 ('in', 1813),
 ('was', 1796),
 ('I', 1740),
 ('that', 1419)]

The number of word  tokens is 124,588.  Counting tokens is
what gets us the raw numbers like 4205 and 4121

In [16]:
list(ctr.values())

[3,
 1,
 29,
 4,
 8,
 1,
 2,
 1,
 5,
 1,
 5,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 2,
 1,
 1,
 1,
 2,
 1,
 4,
 2,
 1,
 1,
 5,
 1,
 1,
 6,
 1,
 1,
 1,
 2,
 4,
 1,
 1,
 2,
 1,
 2,
 1,
 89,
 114,
 1,
 2,
 1,
 18,
 11,
 22,
 1,
 1,
 1,
 1,
 2,
 1,
 3,
 1,
 1,
 1,
 3,
 2,
 1,
 1,
 2,
 1,
 2,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 8,
 3,
 2,
 2,
 1,
 3,
 1,
 1,
 2,
 27,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 8,
 1,
 1,
 2,
 1,
 5,
 1,
 1,
 1,
 1,
 1,
 2,
 2,
 1,
 2,
 1,
 2,
 2,
 2,
 3,
 3,
 1,
 6,
 4,
 3,
 1,
 36,
 1,
 1,
 1,
 14,
 1,
 1,
 31,
 2,
 1,
 1,
 4,
 1,
 3,
 31,
 5,
 3,
 1,
 1,
 1,
 2,
 1,
 2,
 5,
 1,
 1,
 1,
 1,
 8,
 18,
 1,
 1,
 4,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 3,
 1,
 1,
 1,
 1,
 1,
 3,
 1,
 2,
 1,
 1,
 1,
 2,
 19,
 4,
 1,
 2,
 42,
 1,
 1,
 1,
 1,
 3,
 4,
 1,
 1,
 1,
 1,
 1,
 162,
 2,
 2,
 1,
 1,
 1,
 2,
 37,
 1,
 1,
 13,
 1,
 1,
 5,
 3,
 2,
 1,
 1,
 1,
 1,
 2,
 10,
 1,
 1,
 2,
 3,
 2,
 3,
 7,
 1,
 2,
 1,
 1,
 2,
 2,
 1,
 2,
 3,
 1,
 1,
 1,
 3,
 1,
 1,
 73,
 1,
 4,
 12,
 4,
 1,
 2,
 1,
 1,
 2,
 1

In [16]:
sum(ctr.values())

124588

There is an even simpler way to use a counter::

In [17]:
ctr2 = Counter(open('pride_and_prejudice.txt','r'))

In this case however, we iterate through the file-like-object
as in a `for` loop, which means we iterate line by line.
So what the counter will be counting will be lines:

In [18]:
ctr2.most_common(10)

[('\r\n', 2394),
 ('                          * * * * *\r\n', 6),
 ('them."\r\n', 3),
 ('it.\r\n', 3),
 ('them.\r\n', 3),
 ('family.\r\n', 2),
 ('do."\r\n', 2),
 ('between Mr. Darcy and herself.\r\n', 2),
 ('almost no restrictions whatsoever.  You may copy it, give it away or\r\n',
  2),
 ('together.\r\n', 2)]

What would we be counting if we did the following?

In [23]:
ctr3 = Counter(open('pride_and_prejudice.txt','r').read())
ctr3.most_common(10)

[(' ', 113924),
 ('e', 70339),
 ('t', 47283),
 ('a', 42155),
 ('o', 41138),
 ('n', 38428),
 ('i', 36271),
 ('h', 33880),
 ('s', 33292),
 ('r', 33290)]

And now?

In [20]:
ctr4 = Counter(open('pride_and_prejudice.txt','r').read().split())
ctr4.most_common(10)

[('the', 4205),
 ('to', 4121),
 ('of', 3660),
 ('and', 3309),
 ('a', 1945),
 ('her', 1858),
 ('in', 1813),
 ('was', 1796),
 ('I', 1740),
 ('that', 1419)]

In [21]:
ctr4['Elizabeth']

398

In [22]:
ctr4['Lizzy']

13

In [23]:
ctr['Darcy']

215

This is an interesting statistic.  Darcy's mention count is a little over half that of Elizbeths's.
Those who know the book will remember that Elizabeth is in many more scenes
than Darcy.  The fact that Darcy's mention count is over half that of Elizabeth's means that he is being mentioned in
a lot of scenes in which he isn't present.  It would be interesting to know what percentage of "Darcy" mentions
are made in scenes in which he is absent.  And of course it would also be interesting to know the number of scenes 
in which Elizabeth is mentioned but is absent, which is no doubt much lower.  But if you think about it, this is 
going to be a hard number to come up with automatically.  In fact, just segmenting the text into scenes is no trivial
task for a computer,  This is just a very simple example of the kind of information that it is difficult to
extract from text.

## Other types of filelike objects

In [24]:
from collections import Counter
from urllib.request import  urlopen
book_url = 'http://www.gutenberg.org/ebooks/1342.txt.utf-8' # Pride & Prejudice URL
handle = urlopen(book_url)
handle

<addinfourl at 4403238456 whose fp = <socket._fileobject object at 0x1065a91d0>>

In [30]:
next(handle)

'with this eBook or online at www.gutenberg.org\r\n'

In [3]:
next(handle)

'\r\n'

In [4]:
handle.readline()

'This eBook is for the use of anyone anywhere at no cost and with\r\n'

In [31]:
for (i,line) in enumerate(handle):
    if i < 50:
        print(line)
    else:
        break





Title: Pride and Prejudice



Author: Jane Austen



Posting Date: August 26, 2008 [EBook #1342]

Release Date: June, 1998

Last updated: February 15, 2015]



Language: English





*** START OF THIS PROJECT GUTENBERG EBOOK PRIDE AND PREJUDICE ***









Produced by Anonymous Volunteers











PRIDE AND PREJUDICE



By Jane Austen







Chapter 1





It is a truth universally acknowledged, that a single man in possession

of a good fortune, must be in want of a wife.



However little known the feelings or views of such a man may be on his

first entering a neighbourhood, this truth is so well fixed in the minds

of the surrounding families, that he is considered the rightful property

of some one or other of their daughters.



"My dear Mr. Bennet," said his lady to him one day, "have you heard that

Netherfield Park is let at last?"



Mr. Bennet replied that he had not.



"But it is," returned she; "for Mrs. Long has just been here, and she

told me all about it."



M

In [6]:
import zipfile
myzip = zipfile.ZipFile('pride_and_prejudice.zip','r')
with myzip.open('pride_and_prejudice.txt','r') as ifh:
     for (i,line) in enumerate(ifh):
         if i < 50:
            print(line)
         else:
            break

The Project Gutenberg EBook of Pride and Prejudice, by Jane Austen



This eBook is for the use of anyone anywhere at no cost and with

almost no restrictions whatsoever.  You may copy it, give it away or

re-use it under the terms of the Project Gutenberg License included

with this eBook or online at www.gutenberg.org





Title: Pride and Prejudice



Author: Jane Austen



Posting Date: August 26, 2008 [EBook #1342]

Release Date: June, 1998

[Last updated: October 12, 2012]



Language: English





*** START OF THIS PROJECT GUTENBERG EBOOK PRIDE AND PREJUDICE ***









Produced by Anonymous Volunteers











PRIDE AND PREJUDICE



By Jane Austen







Chapter 1





It is a truth universally acknowledged, that a single man in possession

of a good fortune, must be in want of a wife.



However little known the feelings or views of such a man may be on his

first entering a neighbourhood, this truth is so well fixed in the minds

of the surrounding families, that he is co